src/storage/src/source/mod.rs

```diff
@@ -1003,37 +1056,182 @@ where
                         // There were no new messages, check again after a delay
                         (SourceStatus::Alive, MessageProcessing::YieldedWithDelay)
                     }
-                    Ok(NextMessage::Finished) => (SourceStatus::Done, MessageProcessing::Stopped),
+                    Ok(NextMessage::Finished) => (SourceStatus::Done, MessageProcessing::Draining),
                     Err(e) => {
                         output.session(&cap).give(Err(e.to_string()));
                         (SourceStatus::Done, MessageProcessing::Stopped)
                     }
                 };
             }
 
-            bytes_read_counter.inc_by(bytes_read as u64);
-            source_metrics.record_partition_offsets(metric_updates);
+            // Read in any new bindings.  We use these events are the sole indication of whether a
+            // timestamp can be emitted.  Unless we read it back out of the persistent collection,
+            // assume it can be changed.
+            //
+            // XXX(chae): we'll need to change this unless we can have a nb poll (like we do for
+            // reading snapshots).  I think we'll need a nb poll _anyway_ because we'll need to
+            // be able to read in bindings when we lose a CAS write -- this will require a quick
+            // listen OR the ability to read from a range.
+            let listen_events = match futures::executor::block_on(
+                snap_listener.poll_next(Duration::from_secs(2)),
+            ) {
+                Ok(r) => r,
+                // TODO: eventually this will be some sort of structured error?
+                Err(e) => {
+                    if e.to_string().contains("timeout") {
```

Comment:
interesting! @aljoscha and I were just discussing whether any persist users would have some response to timeouts other than "I dunno, panic"